core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/GlobalOrdinalsStringTermsAggregator.java

```diff
@@ -254,109 +294,54 @@ protected final XContentBuilder keyToXContent(XContentBuilder builder) throws IO
      * instead of on the fly for each match.This is beneficial for low cardinality fields, because it can reduce
      * the amount of look-ups significantly.
      */
-    public static class LowCardinality extends GlobalOrdinalsStringTermsAggregator {
+    static class LowCardinality extends GlobalOrdinalsStringTermsAggregator {
 
         private IntArray segmentDocCounts;
-
+        private SortedSetDocValues globalOrds;
         private SortedSetDocValues segmentOrds;
 
-        public LowCardinality(String name, AggregatorFactories factories, ValuesSource.Bytes.WithOrdinals valuesSource,
-                BucketOrder order, DocValueFormat format,
-                BucketCountThresholds bucketCountThresholds, SearchContext context, Aggregator parent,
-                SubAggCollectionMode collectionMode, boolean showTermDocCountError, List<PipelineAggregator> pipelineAggregators,
-                Map<String, Object> metaData) throws IOException {
-            super(name, factories, valuesSource, order, format, bucketCountThresholds, null, context, parent, collectionMode,
-                    showTermDocCountError, pipelineAggregators, metaData);
+        LowCardinality(String name,
+                       AggregatorFactories factories,
+                       ValuesSource.Bytes.WithOrdinals valuesSource,
+                       BucketOrder order,
+                       DocValueFormat format,
+                       BucketCountThresholds bucketCountThresholds,
+                       SearchContext context,
+                       Aggregator parent,
+                       boolean forceDenseMode,
+                       SubAggCollectionMode collectionMode,
+                       boolean showTermDocCountError,
+                       List<PipelineAggregator> pipelineAggregators,
+                       Map<String, Object> metaData) throws IOException {
+            super(name, factories, valuesSource, order, format, bucketCountThresholds, null,
+                context, parent, forceDenseMode, collectionMode, showTermDocCountError, pipelineAggregators, metaData);
             assert factories == null || factories.countAggregators() == 0;
             this.segmentDocCounts = context.bigArrays().newIntArray(1, true);
         }
 
-        // bucketOrd is ord + 1 to avoid a branch to deal with the missing ord
         @Override
-        protected LeafBucketCollector newCollector(final SortedSetDocValues ords,
-                LeafBucketCollector sub) {
-            segmentDocCounts = context.bigArrays().grow(segmentDocCounts, 1 + ords.getValueCount());
+        public LeafBucketCollector getLeafCollector(LeafReaderContext ctx,
+                                                    final LeafBucketCollector sub) throws IOException {
+            if (segmentOrds != null) {
+                mapSegmentCountsToGlobalCounts();
+            }
+            globalOrds = valuesSource.globalOrdinalsValues(ctx);
```

Comment:
It is used in https://github.com/jimczi/elasticsearch/blob/5b79cbe87c5c28cafed068a3910e1298a9dd700d/core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/GlobalOrdinalsStringTermsAggregator.java#L385 to remap the segment ordinals collected in this segment to global ordinals.